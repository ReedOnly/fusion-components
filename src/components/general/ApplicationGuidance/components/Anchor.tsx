import React from 'react';
import {
    AnchorDOMRect,
    OverlayAnchorElement,
    OverlayAnchorElementTag,
    OverlayAnchorElementProps,
    OverlayAnchorConnectEvent,
} from '../../../../customElements/components/overlay/anchor';

export interface AppGuideAnchorRef<R extends HTMLElement> {
    id: string;
    scope?: string;
    padding?: number;
    ref: React.RefObject<R>;
}

export const useAnchor = <R extends HTMLElement>(anchor: Omit<AppGuideAnchorRef<R>, 'ref'>) => {
    const ref = React.useRef<R>(null);
    useAnchorRef({ ...anchor, ref });
    return ref;
};

export const useAnchorRef = <R extends HTMLElement>(anchor: AppGuideAnchorRef<R>) => {
    const { id, ref, scope, padding } = anchor;
    const callBackRef = React.useRef<VoidFunction>();
    const anchorPadding = React.useRef<number>(padding);

    React.useEffect(() => {
        if (!ref.current) {
            return;
        }
        requestAnimationFrame(() => {
            const event = new OverlayAnchorConnectEvent({
                detail: {
                    id,
                    scope,
                    bounds: () => {
                        return AnchorDOMRect.create(
                            ref.current.getBoundingClientRect(),
                            anchorPadding.current
                        );
                    },
                    disconnectedCallback: (cb: VoidFunction) => {
                        callBackRef.current = cb;
                    },
                },
                cancelable: false,
                bubbles: true,
                composed: true,
            });
            ref.current.dispatchEvent(event);
        });
        return () => callBackRef.current();
    }, [ref]);
};

export type ApplicationGuidanceAnchorProps = React.PropsWithChildren<
    Omit<OverlayAnchorElementProps, 'bounds'> &
        React.DetailedHTMLProps<React.HTMLAttributes<OverlayAnchorElement>, OverlayAnchorElement>
>;

export const ApplicationGuidanceAnchor: React.FC<ApplicationGuidanceAnchorProps> = (
    props: ApplicationGuidanceAnchorProps
) => {
    return <fusion-overlay-anchor {...props}></fusion-overlay-anchor>;
};

declare global {
    namespace JSX {
        interface ReactHTML {
            [OverlayAnchorElementTag]: React.DetailedHTMLFactory<
                ApplicationGuidanceAnchorProps,
                OverlayAnchorElement
            >;
        }
        interface IntrinsicElements {
            [OverlayAnchorElementTag]: React.DetailedHTMLProps<
                ApplicationGuidanceAnchorProps,
                OverlayAnchorElement
            >;
        }
    }
}

export default ApplicationGuidanceAnchor;
