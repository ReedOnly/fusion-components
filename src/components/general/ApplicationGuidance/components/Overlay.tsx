import React from 'react';
import {
    OverlayElement,
    OverlayElementProps,
    OverlayElementTag,
    OverlayEvent,
    OverlayEventType,
} from '../../../../customElements/components/overlay';

export type OverlayProps = React.PropsWithChildren<
    OverlayElementProps &
        React.DetailedHTMLProps<React.HTMLAttributes<OverlayElement>, OverlayElement> & {
            onSelection?: (e: OverlayEvent<OverlayEventType.selection>) => void;
            onScope?: (e: OverlayEvent<OverlayEventType.scope>) => void;
        }
>;

declare global {
    namespace JSX {
        interface ReactHTML {
            [OverlayElementTag]: React.DetailedHTMLFactory<OverlayProps, OverlayElement>;
        }
        interface IntrinsicElements {
            [OverlayElementTag]: React.DetailedHTMLProps<OverlayProps, OverlayElement>;
        }
    }
}

export const Overlay: React.FC<OverlayProps> = (args: OverlayProps) => {
    const { scope, onSelection, onScope, ...props } = args;
    const ref = React.useRef<OverlayElement>();
    React.useEffect(() => {
        if (!ref.current) return;
        const el = ref.current;
        el.scope = scope;
        onSelection && el.addEventListener(OverlayEventType.selection, onSelection);
        onScope && el.addEventListener(OverlayEventType.scope, onScope);
        return () => {
            onSelection && el.removeEventListener(OverlayEventType.selection, onSelection);
            onScope && el.removeEventListener(OverlayEventType.scope, onScope);
        };
    }, [ref]);

    React.useEffect(() => {
        if (!ref.current) return;
        ref.current.scope = scope;
    }, [ref, scope]);

    return <fusion-overlay ref={ref} {...props}></fusion-overlay>;
};

export default Overlay;
