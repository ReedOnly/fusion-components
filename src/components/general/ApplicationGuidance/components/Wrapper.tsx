import React from 'react';
import {
    ApplicationGuideElement,
    ApplicationGuideElementProps,
    ApplicationGuideElementTag,
    ApplicationGuideEvent,
    ApplicationGuideEventType,
} from '../../../../customElements/components/application-guide';

export type ApplicationGuidanceWrapperProps = React.PropsWithChildren<
    ApplicationGuideElementProps &
        React.DetailedHTMLProps<
            React.HTMLAttributes<ApplicationGuideElement>,
            ApplicationGuideElement
        > & {
            onOpen?: (e: ApplicationGuideEvent<ApplicationGuideEventType.activated>) => void;
            onClose?: (e: ApplicationGuideEvent<ApplicationGuideEventType.deactivated>) => void;
            onShow?: (e: ApplicationGuideEvent<ApplicationGuideEventType.show>) => void;
        }
>;

declare global {
    namespace JSX {
        interface ReactHTML {
            [ApplicationGuideElementTag]: React.DetailedHTMLFactory<
                ApplicationGuidanceWrapperProps,
                ApplicationGuideElement
            >;
        }
        interface IntrinsicElements {
            [ApplicationGuideElementTag]: React.DetailedHTMLProps<
                ApplicationGuidanceWrapperProps,
                ApplicationGuideElement
            >;
        }
    }
}

export const ApplicationGuidanceWrapper: React.FC<ApplicationGuidanceWrapperProps> = (
    args: ApplicationGuidanceWrapperProps
) => {
    const { onOpen, onClose, onShow, ...props } = args;
    const ref = React.useRef<ApplicationGuideElement>();
    React.useEffect(() => {
        if (!ref.current) return;
        const el = ref.current;
        onOpen && el.addEventListener(ApplicationGuideEventType.activated, onOpen);
        onClose && el.addEventListener(ApplicationGuideEventType.deactivated, onClose);
        onShow && el.addEventListener(ApplicationGuideEventType.show, onShow);
        return () => {
            onOpen && el.removeEventListener(ApplicationGuideEventType.activated, onOpen);
            onClose && el.removeEventListener(ApplicationGuideEventType.deactivated, onClose);
            onShow && el.removeEventListener(ApplicationGuideEventType.show, onShow);
        };
    }, [ref]);
    return <fusion-application-guide ref={ref} {...props}></fusion-application-guide>;
};

export default ApplicationGuidanceWrapper;
