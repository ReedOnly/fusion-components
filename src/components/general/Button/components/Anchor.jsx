import React, { forwardRef } from "react";
import PropTypes from "prop-types";
import Content from "./Content";
import getButtonClasses from "../buttonClasses";

const AnchorComponent = forwardRef(
    (
        {
            children,
            disabled,
            url,
            targetBlank,
            onMouseDown,
            onMouseUp,
            onClick,
            onClickCapture,
            ...props
        },
        ref
    ) => (
        <a
            className={getButtonClasses(props)}
            disabled={disabled}
            href={url}
            target={targetBlank ? "_blank" : "_self"}
            onMouseDown={onMouseDown}
            onMouseUp={onMouseUp}
            onClick={onClick}
            onClickCapture={onClickCapture}
            ref={ref}
        >
            <Content>{children}</Content>
        </a>
    )
);

AnchorComponent.propTypes = {
    children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired,
    disabled: PropTypes.bool,
    url: PropTypes.string.isRequired,
    targetBlank: PropTypes.bool,
    onClick: PropTypes.func,
    onClickCapture: PropTypes.func,
    onMouseDown: PropTypes.func,
    onMouseUp: PropTypes.func,
};

AnchorComponent.defaultProps = {
    disabled: true,
    targetBlank: false,
    onClick: null,
    onClickCapture: null,
    onMouseDown: null,
    onMouseUp: null,
};

AnchorComponent.displayName = "@fusion/components/general/Button/AnchorComponent";

export default AnchorComponent;
