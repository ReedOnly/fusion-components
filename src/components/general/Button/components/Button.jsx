import React, { forwardRef } from "react";
import Content from "./Content";
import getButtonClasses from "../buttonClasses";

const ButtonComponent = forwardRef(
    (
        {
            children,
            disabled,
            onMouseDown,
            onMouseUp,
            onClick,
            onClickCapture,
            ...props
        },
        ref
    ) => (
        <button
            type="button"
            className={getButtonClasses(props)}
            disabled={disabled}
            onMouseDown={onMouseDown}
            onMouseUp={onMouseUp}
            onClick={onClick}
            onClickCapture={onClickCapture}
            ref={ref}
        >
            <Content>{children}</Content>
        </button>
    )
);

ButtonComponent.displayName =
    "@fusion/components/general/Button/ButtonComponent";

export default ButtonComponent;
