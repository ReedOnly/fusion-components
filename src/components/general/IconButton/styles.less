.container {
    background: none;
    border: none;
    outline: none;
    margin: 0;
    padding: 0;
    cursor: pointer;
    overflow: visible;

    .iconContainer {
        display: flex;
        align-items: center;
        justify-content: center;
        width: 100%;
        height: 100%;
        box-sizing: border-box;
        border: 1px solid transparent;
        border-radius: 50%;
        color: var(--color-primary);
    }

    &.isToggler .iconContainer {
        border-radius: 1px;
    }

    &:hover {
        .iconContainer {
             background-color: var(--color-black-alt5);
             border-color: var(--color-black-alt4);
        }
    }

    &.comfortable {
        width: calc(var(--grid-unit) * 6px);
        height: calc(var(--grid-unit) * 6px);
    }

    &.compact {
        width: calc(var(--grid-unit) * 4px);
        height: calc(var(--grid-unit) * 4px);
    }

    &.isActive {
        .iconContainer {
            background-color: var(--color-primary);
            border-color: var(--color-primary);
            color: var(--color-white);
        }
    }

    &:active {
        .iconContainer {
            background-color: var(--color-primary-alt4);
            border-color: var(--color-primary-alt4);
            box-shadow: 0 0 0 2px var(--color-primary-alt4);
            color: var(--color-primary);
        }
    }
}