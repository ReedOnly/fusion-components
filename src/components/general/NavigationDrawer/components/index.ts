export { default as MenuChild } from "./Child";