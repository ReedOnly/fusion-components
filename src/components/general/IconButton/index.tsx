import React from 'react';
import classNames from 'classnames';

import styles from './styles.less';
import { useComponentDisplayType, ComponentDisplayType } from '@equinor/fusion';

type IconButtonProps = {
    active?: boolean;
    toggler?: boolean;
};

const IconButton: React.FC<IconButtonProps> = ({ active, toggler, children }) => {
    const displayType = useComponentDisplayType();
    const buttonClassNames = classNames(styles.container, {
        [styles.compact]: displayType === ComponentDisplayType.Compact,
        [styles.comfortable]: displayType === ComponentDisplayType.Comfortable,
        [styles.isToggler]: toggler,
        [styles.isActive]: toggler && active,
    });

    return (
        <button className={buttonClassNames}>
            <span className={styles.iconContainer}>{children}</span>
        </button>
    );
};

export default IconButton;
