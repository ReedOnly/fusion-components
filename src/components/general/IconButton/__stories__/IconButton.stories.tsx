import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { actions } from '@storybook/addon-actions';
import IconButton from '../index';
import { SearchIcon } from '../../../icons/components/action';

const eventsFromNames = actions('onClick');

storiesOf('General|Icon Button', module)
    .addParameters({ jest: ['IconButton.stories.jsx'] })
    .add('Default', () => (
        <div style={{ padding: 8 }}>
            <IconButton {...eventsFromNames}>
                <SearchIcon />
            </IconButton>
        </div>
    ))
    .add('Toggler', () => (
        <div style={{ padding: 8 }}>
            <IconButton toggler {...eventsFromNames}>
                <SearchIcon />
            </IconButton>
            <IconButton toggler active {...eventsFromNames}>
                <SearchIcon />
            </IconButton>
        </div>
    ));
