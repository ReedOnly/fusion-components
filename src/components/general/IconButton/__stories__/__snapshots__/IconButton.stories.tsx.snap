// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General|Icon Button Default 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      style={
        Object {
          "padding": 8,
        }
      }
    >
      <button
        className="container comfortable"
        onClick={[Function]}
        onMouseDown={[Function]}
      >
        <span
          className="iconContainer"
        >
          <svg
            height={24}
            style={
              Object {
                "color": "",
              }
            }
            viewBox="0 0 24 24"
            width={24}
          >
            <path
              clipRule="evenodd"
              d="M15.1866 14.4716H15.9766L20.2166 18.7316C20.6266 19.1416 20.6266 19.8116 20.2166 20.2216C19.8066 20.6316 19.1366 20.6316 18.7266 20.2216L14.4766 15.9716V15.1816L14.2066 14.9016C12.8066 16.1016 10.8966 16.7216 8.86658 16.3816C6.08658 15.9116 3.86658 13.5916 3.52658 10.7916C3.00658 6.56156 6.56658 3.00156 10.7966 3.52156C13.5966 3.86156 15.9166 6.08156 16.3866 8.86156C16.7266 10.8916 16.1066 12.8016 14.9066 14.2016L15.1866 14.4716ZM5.47658 9.97156C5.47658 12.4616 7.48658 14.4716 9.97658 14.4716C12.4666 14.4716 14.4766 12.4616 14.4766 9.97156C14.4766 7.48156 12.4666 5.47156 9.97658 5.47156C7.48658 5.47156 5.47658 7.48156 5.47658 9.97156Z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;

exports[`General|Icon Button OpenInNew 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      style={
        Object {
          "padding": 8,
        }
      }
    >
      <button
        className="container comfortable"
        onClick={[Function]}
        onMouseDown={[Function]}
      >
        <span
          className="iconContainer"
        >
          <svg
            height={24}
            style={
              Object {
                "color": "",
              }
            }
            viewBox="0 0 24 24"
            width={24}
          >
            <path
              clipRule="evenodd"
              d="M5 18C5 18.55 5.45 19 6 19H18C18.55 19 19 18.55 19 18V13C19 12.45 19.45 12 20 12C20.55 12 21 12.45 21 13V19C21 20.1 20.1 21 19 21H5C3.9 21 3 20.1 3 19V5C3 3.9 3.89 3 5 3H11C11.55 3 12 3.45 12 4C12 4.55 11.55 5 11 5H6C5.45 5 5 5.45 5 6V18ZM15 5C14.45 5 14 4.55 14 4C14 3.45 14.45 3 15 3H20C20.55 3 21 3.45 21 4V9C21 9.55 20.55 10 20 10C19.45 10 19 9.55 19 9V6.41L9.87 15.54C9.48 15.93 8.85 15.93 8.46 15.54C8.07 15.15 8.07 14.52 8.46 14.13L17.59 5H15Z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;

exports[`General|Icon Button Toggler 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      style={
        Object {
          "padding": 8,
        }
      }
    >
      <button
        className="container comfortable isToggler"
        onClick={[Function]}
        onMouseDown={[Function]}
      >
        <span
          className="iconContainer"
        >
          <svg
            height={24}
            style={
              Object {
                "color": "",
              }
            }
            viewBox="0 0 24 24"
            width={24}
          >
            <path
              clipRule="evenodd"
              d="M15.1866 14.4716H15.9766L20.2166 18.7316C20.6266 19.1416 20.6266 19.8116 20.2166 20.2216C19.8066 20.6316 19.1366 20.6316 18.7266 20.2216L14.4766 15.9716V15.1816L14.2066 14.9016C12.8066 16.1016 10.8966 16.7216 8.86658 16.3816C6.08658 15.9116 3.86658 13.5916 3.52658 10.7916C3.00658 6.56156 6.56658 3.00156 10.7966 3.52156C13.5966 3.86156 15.9166 6.08156 16.3866 8.86156C16.7266 10.8916 16.1066 12.8016 14.9066 14.2016L15.1866 14.4716ZM5.47658 9.97156C5.47658 12.4616 7.48658 14.4716 9.97658 14.4716C12.4666 14.4716 14.4766 12.4616 14.4766 9.97156C14.4766 7.48156 12.4666 5.47156 9.97658 5.47156C7.48658 5.47156 5.47658 7.48156 5.47658 9.97156Z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </button>
      <button
        className="container comfortable isToggler isActive"
        onClick={[Function]}
        onMouseDown={[Function]}
      >
        <span
          className="iconContainer"
        >
          <svg
            height={24}
            style={
              Object {
                "color": "",
              }
            }
            viewBox="0 0 24 24"
            width={24}
          >
            <path
              clipRule="evenodd"
              d="M15.1866 14.4716H15.9766L20.2166 18.7316C20.6266 19.1416 20.6266 19.8116 20.2166 20.2216C19.8066 20.6316 19.1366 20.6316 18.7266 20.2216L14.4766 15.9716V15.1816L14.2066 14.9016C12.8066 16.1016 10.8966 16.7216 8.86658 16.3816C6.08658 15.9116 3.86658 13.5916 3.52658 10.7916C3.00658 6.56156 6.56658 3.00156 10.7966 3.52156C13.5966 3.86156 15.9166 6.08156 16.3866 8.86156C16.7266 10.8916 16.1066 12.8016 14.9066 14.2016L15.1866 14.4716ZM5.47658 9.97156C5.47658 12.4616 7.48658 14.4716 9.97658 14.4716C12.4666 14.4716 14.4766 12.4616 14.4766 9.97156C14.4766 7.48156 12.4666 5.47156 9.97658 5.47156C7.48658 5.47156 5.47658 7.48156 5.47658 9.97156Z"
              fill="currentColor"
              fillRule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
